src/Diagnostics.HealthChecks/OmexHealthCheckPublisherHostedService.cs

```diff
@@ -0,0 +1,351 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#pragma warning disable IDE1006 // ASP.NET Core codebase naming convention
+#pragma warning disable IDE0008 // Use explicit type
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks;
+
+internal sealed partial class OmexHealthCheckPublisherHostedService : IHostedService
+{
+	private readonly HealthCheckService _healthCheckService;
+	private readonly IOptions<HealthCheckServiceOptions> _healthCheckServiceOptions;
+    private readonly IOptions<HealthCheckPublisherOptions> _healthCheckPublisherOptions;
+    private readonly IOptions<HealthCheckRegistrationParametersOptions> _healthCheckRegistrationOptions;
+    private readonly ILogger _logger;
+    private readonly IHealthCheckPublisher[] _publishers;
+    private readonly HealthCheckRegistrationParameters _defaultHealthCheckOptions;
+    private readonly Dictionary<HealthCheckRegistrationParameters, List<HealthCheckRegistration>> _healthChecksByOptions;
+    private Dictionary<HealthCheckRegistrationParameters, Timer>? _timersByOptions;
```

Comment:
I'm wondering how expensive running a timer for each health check is and if it wouldn't be better to figure out the schedule between the checks once and the have a single timer sleeping for the deltas.

Reply:
I initially thought about the same. I recall having read the Timer class already reuses a common timer object underneath, so perf impact should be okay-ish (and .NET team doesn't seem worried). By the way the number of timers we create is minimized as we are reusing the same timer for HCs having same delay, period and timeout.